### ISSUE TITLE
materialized_views_test.TestMaterializedViews.test_really_complex_repair mis-interprets the expected result

### ISSUE THREAD

**Issue by xemul** [scylladb/scylladb#15314]
Noticed here: https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/4321/testReport/junit/materialized_views_test/TestMaterializedViews/FullDtest___full_split013___test_really_complex_repair/

```
AssertionError: Expected [['c', 1, 1, 3.0], ['c', 2, 2, 3.0]] from SELECT * FROM ks.t_by_v WHERE v2 = 'c', but got [['c', 1, 1, Decimal('3.0')]]
```

It's mostly matches, except for the `Decimal` cast

**Comment by mykaul** [comment]
@raphaelsc , @nyh - any ideas?

**Comment by nyh** [comment]
> AssertionError: Expected [['c', 1, 1, 3.0], ['c', 2, 2, 3.0]] from SELECT * FROM ks.t_by_v WHERE v2 = 'c', but got [['c', 1, 1, Decimal('3.0')]]
> 
> It's mostly matches, except for the `Decimal` cast

I think the decimal is a red-herring (3.0 is, as far as I know, equal to Decimal('3.0')). I think the problem is that the request expected two rows but just got one. I'll take a look at the test.

**Comment by nyh** [comment]
I believe that this test is buggy, it's not a Scylla bug: 

Since it's an ancient (8 year old) test from Cassandra, I think it's worth looking at the Cassandra dtest to see if it changed, but even without this, I think I know where the bug is:

Basically, this test does the same thing, more or less, multiple times, and fails on the third time, when it is done on a cluster with just 2 out of 5 nodes alive. The keyspace has RF=5 The requests are done with CL=ONE:
```
       session2.execute("INSERT INTO ks.t (id, v, v2, v3) VALUES (1, 1, 'c', 3.0)")
       session2.execute("INSERT INTO ks.t (id, v, v2, v3) VALUES (2, 2, 'c', 3.0)")
       assert_all(session2, "SELECT * FROM ks.t_by_v WHERE v2 = 'c'", [['c', 1, 1, 3.0], ['c', 2, 2, 3.0]])
```

The assert finds just the first row, not the second.

I think I know why:

An "obvious" guess is that the view updates are asynchronous, so it is not surprising that immediately after the second write, it's not yet readable in the view. HOWEVER, we can immediately retort: We have 5 nodes and RF=5, each of the 5 nodes is both a base replica and a view replica, so the view updates are actually guaranteed to be synchronous!

By the story isn't done here... All requests above use CL=ONE. When we do the second INSERT with CL=ONE, it means we wait for just one of the two live nodes to finish their write (both base and view). 

So it's possible that:
1. The second INSERT inserted (the base and the) view data into node A, and will insert it into node B in the background.
2. The SELECT reads from node B (the second live node) and it doesn't yet have the latest update.

If my understanding is correct, the most natural way to fix this test is to replace the assertion by a loop that waits a bit for this to happen eventually. This is very customary in MV tests, and we even have the "eventually" function for it.

**Comment by nyh** [comment]
I see that the Cassandra dtest is identical to what we have, without fixes, but 6 years ago they marked the entire test class TestMaterializedViews as `@flaky` :-)

**Comment by nyh** [comment]
> @nyh - what should be the next step then? At the very least, remove from gating?

This should have been a question for the QA developers, but it sounds like you want me to handle it, so I'll produce a patch (either adding to gating, or better yet, fix the test).

**Comment by nyh** [comment]
As suspected, this test is flaky when I run it locally with the command:
> PYTHONPATH=../scylla-ccm pytest --cassandra-dir=$HOME/scylla/build/dev materialized_views_test.py::TestMaterializedViews::test_really_complex_repair

It failed just 1 time out of 5 runs (the test takes 4 minutes, so I didn't try more).

I'm trying to think how this test can be "salvaged" - be made correct and still test whatever it intended to test. I don't want to just remove it from "gating" - that's not a good solution, we will see its failures once in a while outside gating :-(

Another option besides really fixing it is adding a "flaky" decorator, which will cause the test to be retried if it fails just once. I see we the have support for the `@flaky` tag in dtest, but it's not used anywhere... @fruch, is it supposed to work?

**Comment by mykaul** [comment]
I don't think we've managed to get an agreement on the flaky tag behavior. I'd be happy to see it in, in a non-voting manner. Others are less excited by this.

**Comment by nyh** [comment]
> I don't think we've managed to get an agreement on the flaky tag behavior. I'd be happy to see it in, in a non-voting manner. Others are less excited by this.

I'm not a fan of the "flaky" tag, it can cause us to be lazy and not fix buggy tests. In this case I prefer the fix the test. Maybe "flaky" is still useful as some sort of emergency situation where we can't afford to waste time on the test and just want to leave it to the future. But it shouldn't be the first line of defense.

**Comment by fruch** [comment]
> > I don't think we've managed to get an agreement on the flaky tag behavior. I'd be happy to see it in, in a non-voting manner. Others are less excited by this.
> 
> I'm not a fan of the "flaky" tag, it can cause us to be lazy and not fix buggy tests. In this case I prefer to fix the buggy test. Maybe "flaky" is still useful as some sort of emergency situation where we can't afford to waste time on the test and just want to leave it to the future. But it shouldn't be the first line of defense.

our emergency case right now, it to take it from gating, it would still run daily, and we can continue the investigation without blocking next/PRs, but putting flaky into the mix, won't get us to stable tests, if it fail 1/3, sometimes it would fail 3/3.
anyhow marking test flaky and gating, doesn't sound like something we'll do anyhow.

@nyh as for the technical question, we do have https://pypi.org/project/flaky/ available, but we run the tests with `-p no:flaky` 
so using `@flaky` won't be working since https://github.com/scylladb/scylla-dtest/pull/2599

**Comment by mykaul** [comment]
I wonder why the bot did not add the backport label here. In any case, I've asked that the dtest fix will be backported to 5.2.
